pydantic_ai_slim/pydantic_ai/mcp.py

```diff
@@ -50,6 +59,166 @@
+
+        self.wrapped = wrapped
+
+        # If it's a CombinedToolset, filter out mcp-run-python toolsets before any get_tools() calls
+        if isinstance(wrapped, CombinedToolset):
+            filtered_toolsets: list[AbstractToolset[Any]] = []
+            for toolset in wrapped.toolsets:
+                if not _is_mcp_run_python_server(toolset):
+                    filtered_toolsets.append(toolset)
+            # Create a new CombinedToolset with filtered toolsets
+            self.wrapped = CombinedToolset(toolsets=filtered_toolsets)
+        elif _is_mcp_run_python_server(wrapped):
+            # If the wrapped toolset itself is mcp-run-python, return empty tools
+            self._is_empty = True
+        else:
+            self._is_empty = False
+
+    async def get_tools(self, ctx: RunContext[Any]) -> dict[str, ToolsetTool[Any]]:
+        """Get tools excluding those from mcp-run-python servers."""
+        if hasattr(self, '_is_empty') and self._is_empty:
+            return {}
+
+        return await self.wrapped.get_tools(ctx)
+
+    async def call_tool(
+        self, name: str, tool_args: dict[str, Any], ctx: RunContext[Any], tool: ToolsetTool[Any]
+    ) -> Any:
+        """Call tool using the wrapped toolset."""
+        return await self.wrapped.call_tool(name, tool_args, ctx, tool)
+
+    def apply(self, visitor: Callable[[AbstractToolset[Any]], None]) -> None:
+        """Apply visitor to the wrapped toolset."""
+        self.wrapped.apply(visitor)
+
+
+async def _call_tool_for_elicitation(
+    toolset: AbstractToolset[Any],
+    tool_name: str,
+    tool_arguments: dict[str, Any],
+    ctx: RunContext[Any],
+) -> Any:
+    """Call a tool for elicitation using standard toolset API."""
+    available_tools = await toolset.get_tools(ctx)
+    if tool_name not in available_tools:
+        raise ValueError(f'Tool {tool_name} not found in toolset')
+
+    tool = available_tools[tool_name]
+    return await toolset.call_tool(tool_name, tool_arguments, ctx, tool)
+
+
+def _create_elicitation_context(tool_name: str) -> RunContext[Any]:
+    """Create minimal RunContext for elicitation."""
+    return RunContext(
+        deps=None,
+        model=TestModel(),
+        usage=Usage(),
+        tool_name=tool_name,
+    )
+
+
+def create_tool_elicitation_callback(toolset: AbstractToolset[Any]) -> ElicitationFnT:
+    """Create an elicitation callback that routes to the provided toolset.
+
+    Args:
+        toolset: The toolset containing tools that can be called via elicitation
+
+    Returns:
+        An elicitation callback function that handles tool execution requests
+    """
+    filtered_toolset = _FilteredToolset(toolset)
+
+    async def elicitation_callback(context: Any, params: Any) -> mcp_types.ElicitResult | mcp_types.ErrorData:
+        """Handle elicitation requests by routing to the toolset."""
+        try:
+            tool_execution_data = json.loads(params.message)
```

Comment:
Can we use `TypeAdapter(ToolCallPart)` to validate this, which we can then directly pass through to `tool_manager.handle_call` (see my other suggestion about using that instead of `call_tool`)?